File: Java/src/main/java/gcs/WriteTextIO.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gcs;

import com.google.common.primitives.Ints;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.TextIO;
import org.apache.beam.sdk.options.Default;
import org.apache.beam.sdk.options.Description;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.transforms.Create;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.ParDo;

import java.util.stream.IntStream;

public class WriteTextIO {

    public interface writeTextIOOptions extends PipelineOptions {
        @Description("Output to write to")
        String getOutput();

        void setOutput(String value);

        @Description("Max Number")
        @Default.Integer(100)
        Integer getMaxNumber();

        void setMaxNumber(Integer value);
    }

    public static void main(String[] args) {
        // Reference the extended class
        writeTextIOOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(writeTextIOOptions.class);

        Pipeline p = Pipeline.create(options);

        int[] rangeIntegers = IntStream.range(0, options.getMaxNumber()).toArray();
        Iterable<Integer> elements = Ints.asList(rangeIntegers);

        p
                .apply(Create.of(elements))
                .apply(ParDo.of(new DoFn<Integer, String>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        c.output("This is element number " + c.element().toString());
                    }
                }))
                .apply("Write File(s)",
                        TextIO.write()
                                .to(options.getOutput()));

        p.run();


    }

}

File: Python/gcs/write_textio.py
#  Copyright 2022 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# standard libraries
import logging

# third party libraries
import apache_beam as beam
from apache_beam import Create, Map
from apache_beam.io.textio import WriteToText
from apache_beam.options.pipeline_options import PipelineOptions


def run(argv=None):
    class WriteTextOptions(PipelineOptions):
        @classmethod
        def _add_argparse_args(cls, parser):
            parser.add_argument(
                "--output",
                description="Output to write to",
                dest="output",
                required=True,
                help="Output file to write results to.",
            )
            parser.add_argument(
                "--max_number",
                dest="max_number",
                default=100,
                help="Number of shards to process.",
            )

    options = WriteTextOptions()
    elements = range(options.max_number)

    with beam.Pipeline(options=options) as p:
        (
            p
            | Create(elements)
            | "Format String"
            >> Map(
                lambda x: f"This element number is {x}"
            )  # Changing number to string and
            | "Write Files" >> WriteToText(options.output)
        )


if __name__ == "__main__":
    logging.getLogger().setLevel(logging.INFO)
    run()
